### ISSUE TITLE
Should we have subobject provenance?

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/unsafe-code-guidelines#419]
Tree Borrows fixes https://github.com/rust-lang/unsafe-code-guidelines/issues/256. On the flip side, Tree Borrows doesn't enforce subobject provenance: when you create an `&mut x.field`, you can use it to mutate other fields, as long as uniqueness is maintained. We could try to come up with something clever that both supports `&Header` and enforces subobject provenance but I am not convinced that that's worth it. So far the only proposal I saw is to make this depend on whether the type of `x` is `repr(C)` and I find that way too confusing.

**Comment by joboet** [comment]
I think there are some very valuable optimisations that could be enabled by subobject provenance (once LLVM adds support for it). For instance, consider
```rust
let rc = Rc::new(42);
takes_ref(&*rc);
```
The `Rc` allocation [cannot be removed](https://godbolt.org/z/bE7rq39jT) if `takes_ref` is not inlined, as the optimiser cannot prove that `takes_ref` does not access the reference counts. Subobject provenance however allows optimising this to just `takes_ref(const { &42 })`.

For another cool example consider
```rust
let file = File::open("./some_rust_string.txt");
```
On UNIX, this will end up calling `run_with_cstr_stack`, which will copy the string into a fixed-size stack buffer:

https://github.com/rust-lang/rust/blob/e1b06f773063140263558b858cb6e4c99755a763/library/std/src/sys/pal/common/small_c_string.rs#L32-L51

Ideally, the optimiser would eliminate this copy and just make the string NUL-terminated. But this is only possible if the provenance of the `CStr` is constrained to just the initialised bytes of the buffer, as otherwise the optimiser has no way of knowing that the other, unused bytes in the buffer are not accessed by the underlying filesystem API.

As for the `&Header` case, I think the best way to resolve that is by not constraining the provenance of references to extern types, as in those cases the size is truly not known.

**Comment by RalfJung** [comment]
Yeah there definitely are valuable optimizations, but it's also a big footgun and the number 1 or 2 reason for code being UB wrt Stacked Borrows. So the question is whether the optimization is valuable enough to justify all the pain that UB will cause, and whether we can find some trade-off that reduces the chance of people getting this wrong.

For `Rc` specifically, there are also users that want to reconstruct an `Rc<T>` from an `&T`; that can be quite useful and having to carefully only use raw pointers for that is painful -- so it's far from clear that we'd want to apply subobject provenance to `Rc`.

**Comment by Jules-Bertholet** [comment]
Subobject provenance is also necessary (though not sufficient) to resolve a major issue with variadic generic function parameters: https://rust-lang.zulipchat.com/#narrow/channel/136281-t-opsem/topic/Varargs.20SROA.20sadness

That being said, I agree that adopting it everywhere would be too disruptive. I hope we can find a workable compromise. Maybe restrict it to `repr(Rust)` non-ZSTs, at least by default?

**Comment by RalfJung** [comment]
I propose we commit to *not* having subobject provenance for arrays/slices: https://github.com/rust-lang/unsafe-code-guidelines/issues/622.

**Comment by CAD97** [comment]
> We could try to come up with something clever that both supports `&Header` and enforces subobject provenance

A simple option would be to make it a question of whether `Header`'s unsize-tail is an extern type (i.e. if `Header: MetaSized` holds). It's not a particularly nice option to existing code, but it's *an* option. It would probably also only be useful with some simple way to turn a place into an extern-type-tailed version of the same place.

**Comment by RalfJung** [comment]
I think we should forbid extern type fields except maybe in repr(transparent) types -- we currently have the situation where `struct.field` can panic (because an extern type field does not have an alignment and therefore its offset is unknown) which is not good. That makes them unsuited for this task.

This also doesn't help for all the existing code that does pointer arithmetic in arrays/slices.
